### PR TITLE
Recursive Create

### LINKED ISSUE #395: Fix stRecursiveCreate from EF

### LINKED ISSUE #402: Fix stCreateTest from EF

### LINKED ISSUE #405: Fix stCreate2 from EF

### LINKED ISSUE #412: Fix stRandom from EF

### PR DESCRIPTION
**Pull request by belfortep**
closes #395 
closes #402 
closes #405 
closes #412 
## Summary
This PR fix some create issues and add the possibility to have transactions of type `Create`

## Changes
- Fix stCreate, stCreate2, stRecursiveCreate, stRandom and stRandom2 from the EF tests
- Separated the logic of `transact_commit` depending if it's a `Call` or a `Create` transaction

### CODE REVIEW

File: tests/evm.rs [hunk 1]
```diff
@@ -5641,3 +5642,77 @@ fn refund_limit_value() {
...
+        Operation::Push((1, BigUint::ZERO)),
+        Operation::Push((1, BigUint::ZERO)),
+        Operation::Create,
+        Operation::Stop,
+    ];
+
+    let mut env = Env::default();
+    env.tx.value = EU256::from(value);
+    env.tx.caller = sender_addr;
+    env.tx.transact_to = TransactTo::Call(Address::from_low_u64_be(3000));
```
Comment (juanimedone) [inline tests/evm.rs:5668]: Just a detail:
```suggestion
    env.tx.transact_to = TransactTo::Call(to_addr);
```

File: tests/evm.rs [hunk 2]
```diff
@@ -5641,3 +5642,77 @@ fn refund_limit_value() {
...
+    let to_addr = Address::from_low_u64_be(3000);
+
+    let operations = vec![
+        Operation::Push((1, BigUint::from(32_u8))),
+        Operation::Push((1, BigUint::ZERO)),
+        Operation::Push((1, BigUint::ZERO)),
+        Operation::Codecopy,
+        Operation::Push((1, BigUint::from(32_u8))),
+        Operation::Push((1, BigUint::ZERO)),
+        Operation::Push((1, BigUint::ZERO)),
```
Comment (juanimedone) [inline tests/evm.rs:5660]: We have a specific opcode for pushing zeros:
```suggestion
        Operation::Push0,
        Operation::Push0,
        Operation::Codecopy,
        Operation::Push((1, BigUint::from(32_u8))),
        Operation::Push0,
        Operation::Push0,
```

Comment (belfortep) [reply tests/evm.rs:5660]: Ah you are right, changing them

File: src/syscall.rs [hunk 3]
```diff
@@ -866,9 +882,23 @@ impl<'c> SyscallContext<'c> {
...
+
+        self.inner_context.resize_memory_if_necessary(offset, size);
+
         let initialization_bytecode = &self.inner_context.memory[offset..offset + size];
         let program = Program::from_bytecode(initialization_bytecode);
 
+        // if we do a create from a program, and the created program would be the same, that means a recursive create
+        // and we should directly halt to avoid an stack overflow
+        if self.inner_context.program == program.clone().to_bytecode() {
+            self.halt_reason = Some(HaltReason::OutOfGas(crate::result::OutOfGasError::Basic));
```
Comment (juanimedone) [inline src/syscall.rs:898]: We could add a new `HaltReason::RecursiveCreate` or `OutOfGasError::RecursiveCreate` here.

File: src/lib.rs [hunk 4]
```diff
@@ -46,38 +47,85 @@ impl<DB: Database + Default> Evm<DB> {
...
+
+        let mut value = self.get_env_value();
+        let mut remaining_gas = self.env.tx.gas_limit;
+        let gas_limit = self.env.tx.gas_limit;
+        let program = self.env.tx.data.to_vec();
+        let program_size = program.len() as u32;
+        let mut context = self.create_syscall_context();
+        context.inner_context.memory = program;
+
+        context.create(program_size, 0, &mut value, &mut remaining_gas);
```
[the remark this answers is not in the document]
Comment (belfortep) [reply src/lib.rs:116]: Yes, the initial code in `create` transactions is loaded to the inner_context memory starting from there, should we add a constant for more clarity?

File: src/lib.rs [hunk 5]
```diff
@@ -46,38 +47,85 @@ impl<DB: Database + Default> Evm<DB> {
...
+        context.inner_context.gas_remaining = Some(gas_limit.saturating_sub(remaining_gas));
+        context.get_result()
+    }
+
+    /// Executes [the configured transaction](Env::tx).
+    pub fn transact(&mut self) -> Result<ResultAndState, EVMError> {
```
[the remark this answers is not in the document]
Comment (belfortep) [reply src/lib.rs:122]: Ah you are right, changing that

File: src/syscall.rs (hunk 3 above)
Comment (belfortep) [reply src/syscall.rs:899]: I like the `OutOfGasError::RecursiveCreate`, adding that

File: src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (juanimedone) [reply src/lib.rs:113]: Oh ok, it's fine like this.

**Review by juanimedone — APPROVED** [review]

File: tests/evm.rs [hunk 6]
```diff
@@ -124,6 +125,7 @@ fn run_program_assert_gas_and_refund(
     let used_gas = used_gas + env.calculate_intrinsic_cost();
     let mut evm = Evm::new(env, db);
     let result = evm.transact_commit().unwrap();
+    eprintln!("RESULT ES: {:?}", result);
```
Comment (juanimedone) [inline tests/evm.rs:128]:
```suggestion
```

File: src/env.rs [hunk 7]
```diff
@@ -161,6 +162,8 @@ pub struct TxEnv {
     pub caller: Address,
     /// The gas limit of the transaction.
     pub gas_limit: u64,
+    /// The initial gas of the transaction.
+    pub initial_gas: u64,
```
Comment (TomasArrachea) [inline src/env.rs:166]: `TxEvn` is a public struct shared with revm's public API, we should aim to keep it exactly the same for compatibility. Can this new attribute be avoided?

Comment (belfortep) [reply src/env.rs:166]: Oh okay, I think we can avoid the `initial_gas` calculation for another PR then, as it's a little tricky without adding this field and the tests of `stCreate` works either way

File: src/env.rs [hunk 8]
```diff
@@ -26,10 +26,10 @@ pub struct Env {
 }
 
 impl Env {
-    pub fn consume_intrinsic_cost(&mut self) -> Result<(), InvalidTransaction> {
+    pub fn consume_intrinsic_cost(&mut self) -> Result<u64, InvalidTransaction> {
         if self.tx.gas_limit >= self.calculate_intrinsic_cost() {
             self.tx.gas_limit -= self.calculate_intrinsic_cost();
-            Ok(())
+            Ok(self.calculate_intrinsic_cost())
```
Comment (TomasArrachea) [inline src/env.rs:32]: we could move the `self.calculate_intrinsic_cost()` to a separate variable to compute it only once

**Review by TomasArrachea — APPROVED** [review]
LGTM

**Pull request merged**